File: stack and queue/Next Greater Element I.java
import java.util.*;
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int n = nums2.length;
        int[] nextGreater = new int[n];
        Deque<Integer> linearlyDecreasing = new LinkedList<>();

        for(int i = n - 1; i >= 0; i--){
            int curr = nums2[i];
            while(linearlyDecreasing.size() != 0 && curr > linearlyDecreasing.peekLast()){
                linearlyDecreasing.removeLast();
            }

            if(linearlyDecreasing.size() == 0){
                nextGreater[i] = -1;
            }
            else{
                nextGreater[i] = linearlyDecreasing.peekLast();
            }
            linearlyDecreasing.add(curr);
        }

        int m = nums1.length;
        int ans[] = new int[m];
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(nums2[j] == nums1[i]){
                    ans[i] = nextGreater[j];
                    break;
                }
            }
        }
        return ans;
    }
}

File: stack and queue/Next Greater Element I.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int> increasing;
        int n = nums2.size();
        unordered_map<int, int> mp;

        for(int i = n - 1; i >= 0; i--){
            while(!increasing.empty() && nums2[i] > increasing.top()){
                increasing.pop();
            }

            if(increasing.empty()){
                mp[nums2[i]] = -1;
            }
            else{
                mp[nums2[i]] = increasing.top();
            }
            
            increasing.push(nums2[i]);
        }

        int m = nums1.size();
        vector<int> ans;
        for(int i = 0; i < m; i++){
            ans.push_back(mp[nums1[i]]);
        }
        return ans;
    }
};